fireant/queries/builder/dataset_blender_query_builder/klass.py

```diff
@@ -0,0 +1,387 @@
+        as the join criterion for the provided sub-query.
+
+        :param primary_table_sub_query:
+            a PyPika Query instance.
+        :param dataset:
+            A dataset instance.
+        :param sub_query:
+            A PyPika Query instance.
+        :param dimensions:
+            A list of Field instances.
+        :return: a PyPika boolean expression.
+        """
+        criterion = None
+        dimensions_without_mapping = []
+        field_mapping = self.dataset.field_mapping[dataset.table]
+
+        for dimension in dimensions:
+            primary_field = dimension if isinstance(dimension, Field) else dimension.dimension
+            try:
+                secondary_field = field_mapping[primary_field]
+            except KeyError:
+                dimensions_without_mapping.append(dimension)
+                continue
+
+            primary_pypika_field = getattr(
+                primary_table_sub_query, alias_selector(primary_field.alias),
+            )
+            secondary_pypika_field = getattr(sub_query, alias_selector(secondary_field.alias))
+
+            if criterion is None:
+                criterion = primary_pypika_field == secondary_pypika_field
+            else:
+                criterion &= primary_pypika_field == secondary_pypika_field
+
+        return criterion, dimensions_without_mapping
+
+    @property
+    def __queries_per_dataset(self):
+        """
+        Returns a list of Pypika Query instances with dimensions, metrics, filters and references applied.
+
+        :return: a list of DataSetQuery instances with dimensions, metrics, filters and references applied.
+        """
+        queries_per_dataset = {}
+
+        for dataset in self.all_datasets:
+            queries_per_dataset[dataset.table] = dataset.query
+
+        dimensions_per_dataset = self.__dimensions_per_dataset
+        metrics_per_dataset = self.__metrics_per_dataset
+        filters_per_dataset = self.__filters_per_dataset
+        references_per_dataset = self.__references_per_dataset
+
+        for dataset, dataset_dimensions in dimensions_per_dataset.items():
+            for dimension in dataset_dimensions:
+                queries_per_dataset[dataset.table] = queries_per_dataset[dataset.table].dimension(dimension)
+
+        for dataset, dataset_metrics in metrics_per_dataset.items():
+            if not dataset_metrics:
+                continue
+
+            new_widgets = [copy.deepcopy(widget) for widget in self._widgets]
+            for widget in new_widgets:
+                widget.items = [copy.deepcopy(metric) for metric in dataset_metrics]
+                queries_per_dataset[dataset.table] = queries_per_dataset[dataset.table].widget(widget)
+
+        for dataset, dataset_filters in filters_per_dataset.items():
+            for filter in dataset_filters:
+                new_filter = copy.deepcopy(filter)
+                new_definition = copy.deepcopy(filter.definition)
```

Comment:
deepcopy should create a new instance of the definition, no? so this second deepcopy is redundant?

Reply:
I'll remove it.